File: backend_express/src/routes/metrics.ts
import { Router } from 'express'
import { pool } from '../db/index.js'
import { authMiddleware, type AuthRequest } from '../middleware/auth.js'

export const metricsRouter = Router()

async function handleMetrics(req: AuthRequest, res: import('express').Response) {
  const userId = req.user!.id
  const client = await pool.connect()
  try {
    const totalJobsRes = await client.query(
      `SELECT COUNT(j.id)::int AS count
       FROM jobs j
       JOIN videos v ON j.video_id = v.id
       WHERE v.usuario_id = $1`,
      [userId]
    )
    const total_jobs: number = totalJobsRes.rows[0]?.count ?? 0

    const totalClipsRes = await client.query(
      `SELECT COUNT(c.id)::int AS count
       FROM clips c
       JOIN jobs j ON c.job_id = j.id
       JOIN videos v ON j.video_id = v.id
       WHERE v.usuario_id = $1`,
      [userId]
    )
    const total_clips: number = totalClipsRes.rows[0]?.count ?? 0

    const minutesRes = await client.query(
      `SELECT COALESCE(SUM(v.duration_seconds), 0) AS total_seconds
       FROM videos v
       WHERE v.usuario_id = $1`,
      [userId]
    )
    const totalSeconds = Number(minutesRes.rows[0]?.total_seconds ?? 0)
    const total_minutes_processed: number = Math.round((totalSeconds / 60) * 100) / 100

    let time_saved_hours: number
    if (total_minutes_processed > 0) {
      time_saved_hours = Math.round((total_minutes_processed * 3 / 60) * 100) / 100
    } else {
      time_saved_hours = Math.round((total_clips * 15 / 60) * 100) / 100
    }

    const platformRes = await client.query(
      `SELECT c.social_network, COUNT(c.id)::int AS count
       FROM clips c
       JOIN jobs j ON c.job_id = j.id
       JOIN videos v ON j.video_id = v.id
       WHERE v.usuario_id = $1
       GROUP BY c.social_network`,
      [userId]
    )
    const platform_distribution: Record<string, number> = {
      tiktok: 0,
      youtube: 0,
      instagram: 0,
    }
    for (const row of platformRes.rows as Array<{ social_network: string | null; count: number }>) {
      if (!row.social_network) continue
      const s = String(row.social_network).toLowerCase()
      const c = Number(row.count)
      if (s === 'tiktok') platform_distribution.tiktok += c
      else if (s === 'youtube' || s === 'youtube_shorts') platform_distribution.youtube += c
      else if (s === 'instagram' || s === 'instagram_reels') platform_distribution.instagram += c
      else platform_distribution[s] = (platform_distribution[s] ?? 0) + c
    }

    const today = new Date()
    const todayUtc = new Date(Date.UTC(today.getUTCFullYear(), today.getUTCMonth(), today.getUTCDate()))
    const startDate = new Date(todayUtc)
    startDate.setUTCDate(todayUtc.getUTCDate() - 6)
    const startDateStr = startDate.toISOString().slice(0, 10)

    const jobsByDayRes = await client.query(
      `SELECT (j.created_at::date)::text AS d, COUNT(j.id)::int AS count
       FROM jobs j
       JOIN videos v ON j.video_id = v.id
       WHERE v.usuario_id = $1 AND (j.created_at::date) >= $2::date
       GROUP BY d`,
      [userId, startDateStr]
    )
    const jobsMap = new Map<string, number>()
    for (const r of jobsByDayRes.rows as Array<{ d: string; count: number }>) {
      jobsMap.set(r.d, Number(r.count))
    }

    const clipsByDayRes = await client.query(
      `SELECT (c.created_at::date)::text AS d, COUNT(c.id)::int AS count
       FROM clips c
       JOIN jobs j ON c.job_id = j.id
       JOIN videos v ON j.video_id = v.id
       WHERE v.usuario_id = $1 AND (c.created_at::date) >= $2::date
       GROUP BY d`,
      [userId, startDateStr]
    )
    const clipsMap = new Map<string, number>()
    for (const r of clipsByDayRes.rows as Array<{ d: string; count: number }>) {
      clipsMap.set(r.d, Number(r.count))
    }

    const minutesByDayRes = await client.query(
      `SELECT (v.created_at::date)::text AS d, COALESCE(SUM(v.duration_seconds),0) AS total_seconds
       FROM videos v
       WHERE v.usuario_id = $1 AND (v.created_at::date) >= $2::date
       GROUP BY d`,
      [userId, startDateStr]
    )
    const minutesMap = new Map<string, number>()
    for (const r of minutesByDayRes.rows as Array<{ d: string; total_seconds: string }>) {
      const mins = Math.round((Number(r.total_seconds) / 60) * 100) / 100
      minutesMap.set(r.d, mins)
    }

    const recent_activity: Array<{ date: string; jobs: number; clips: number; minutes_processed: number }> = []
    for (let i = 0; i < 7; i++) {
      const d = new Date(startDate)
      d.setUTCDate(startDate.getUTCDate() + i)
      const dStr = d.toISOString().slice(0, 10)
      recent_activity.push({
        date: dStr,
        jobs: jobsMap.get(dStr) ?? 0,
        clips: clipsMap.get(dStr) ?? 0,
        minutes_processed: minutesMap.get(dStr) ?? 0,
      })
    }

    return res.json({
      total_jobs,
      total_clips,
      total_minutes_processed,
      time_saved_hours,
      platform_distribution,
      recent_activity,
    })
  } catch (err) {
    console.error('GET /metrics error', err)
    return res.status(500).json({ detail: 'Error interno' })
  } finally {
    client.release()
  }
}

metricsRouter.get('/metrics', authMiddleware, handleMetrics)
metricsRouter.get('/api/metrics', authMiddleware, handleMetrics)

File: backend_fastapi/app/routers/metrics.py
"""Router de métricas — GET /metrics (Issue 19)."""

from __future__ import annotations

from datetime import datetime, timedelta, timezone

from fastapi import APIRouter
from sqlalchemy import func, select

from ..deps import CurrentUser, DbSession
from ..models import Clip, Job, Video
from ..schemas.metrics import MetricsResponse, RecentActivityItem

router = APIRouter(tags=["metrics"])


@router.get(
    "/metrics",
    response_model=MetricsResponse,
    summary="Métricas del usuario autenticado (dashboard)",
)
@router.get(
    "/api/metrics",
    response_model=MetricsResponse,
    summary="Métricas del usuario autenticado (alias)",
    include_in_schema=False,
)
def get_metrics(
    current_user: CurrentUser,
    db: DbSession,
) -> MetricsResponse:
    user_id = current_user.id

    total_jobs: int = db.scalar(
        select(func.count(Job.id))
        .select_from(Job)
        .join(Video, Job.video_id == Video.id)
        .where(Video.user_id == user_id)
    ) or 0

    total_clips: int = db.scalar(
        select(func.count(Clip.id))
        .select_from(Clip)
        .join(Job, Clip.job_id == Job.id)
        .join(Video, Job.video_id == Video.id)
        .where(Video.user_id == user_id)
    ) or 0

    total_seconds: float = db.scalar(
        select(func.coalesce(func.sum(Video.duration_seconds), 0)).where(Video.user_id == user_id)
    ) or 0
    total_minutes_processed: float = round(float(total_seconds) / 60.0, 2)

    if total_minutes_processed > 0:
        time_saved_hours: float = round(total_minutes_processed * 3 / 60.0, 2)
    else:
        time_saved_hours = round(total_clips * 15 / 60.0, 2)

    platform_rows = db.execute(
        select(Clip.social_network, func.count(Clip.id))
        .select_from(Clip)
        .join(Job, Clip.job_id == Job.id)
        .join(Video, Job.video_id == Video.id)
        .where(Video.user_id == user_id)
        .group_by(Clip.social_network)
    ).all()

    platform_distribution: dict[str, int] = {"tiktok": 0, "youtube": 0, "instagram": 0}
    for social, cnt in platform_rows:
        if social is None:
            continue
        s = str(social).lower()
        c = int(cnt)
        if s == "tiktok":
            platform_distribution["tiktok"] += c
        elif s in ("youtube", "youtube_shorts"):
            platform_distribution["youtube"] += c
        elif s in ("instagram", "instagram_reels"):
            platform_distribution["instagram"] += c
        else:
            platform_distribution[s] = platform_distribution.get(s, 0) + c

    today = datetime.now(timezone.utc).date()
    start_date = today - timedelta(days=6)

    jobs_by_day_rows = db.execute(
        select(func.date(Job.created_at), func.count(Job.id))
        .select_from(Job)
        .join(Video, Job.video_id == Video.id)
        .where(Video.user_id == user_id, func.date(Job.created_at) >= start_date)
        .group_by(func.date(Job.created_at))
    ).all()
    jobs_map: dict[str, int] = {str(r[0]): int(r[1]) for r in jobs_by_day_rows if r[0] is not None}

    clips_by_day_rows = db.execute(
        select(func.date(Clip.created_at), func.count(Clip.id))
        .select_from(Clip)
        .join(Job, Clip.job_id == Job.id)
        .join(Video, Job.video_id == Video.id)
        .where(Video.user_id == user_id, func.date(Clip.created_at) >= start_date)
        .group_by(func.date(Clip.created_at))
    ).all()
    clips_map: dict[str, int] = {str(r[0]): int(r[1]) for r in clips_by_day_rows if r[0] is not None}

    minutes_by_day_rows = db.execute(
        select(func.date(Video.created_at), func.coalesce(func.sum(Video.duration_seconds), 0))
        .where(Video.user_id == user_id, func.date(Video.created_at) >= start_date)
        .group_by(func.date(Video.created_at))
    ).all()
    minutes_map: dict[str, float] = {
        str(r[0]): round(float(r[1]) / 60.0, 2) for r in minutes_by_day_rows if r[0] is not None
    }

    recent_activity: list[RecentActivityItem] = []
    for i in range(7):
        d = start_date + timedelta(days=i)
        d_str = d.isoformat()
        recent_activity.append(
            RecentActivityItem(
                date=d_str,
                jobs=int(jobs_map.get(d_str, 0)),
                clips=int(clips_map.get(d_str, 0)),
                minutes_processed=float(minutes_map.get(d_str, 0.0)),
            )
        )

    return MetricsResponse(
        total_jobs=int(total_jobs),
        total_clips=int(total_clips),
        total_minutes_processed=float(total_minutes_processed),
        time_saved_hours=float(time_saved_hours),
        platform_distribution=platform_distribution,
        recent_activity=recent_activity,
    )
